### PR TITLE
bakery: more intelligent prefix handling for PublicKeyRing

### PR DESCRIPTION
**Pull request by rogpeppe**
The previous string-prefix-based handling was seriously flawed.

### CODE REVIEW

File: bakery/keys.go [hunk 1]
```diff
@@ -148,41 +148,77 @@ func NewPublicKeyRing() *PublicKeyRing {
 }
 
 // AddPublicKeyForLocation adds a public key to the keyring for the given
-// location or location prefix.
+// location. If prefix is true, then inexact locations will be allowed
+// (see PublicKeyForLocation). The matching is similar to that
+// of http.ServeMux, For example, http://foo.com/x/ matches http://foo.com/x/y
+// but http://foo.com/x does not.
+// The method is not significant.
```
Comment (rogpeppe) [inline bakery/keys.go]: ah, i should say "scheme" here

**Comment by mhilton** [comment]
LGTM

**Comment by mhilton** [comment]
might need a gofmt though.

File: bakery/keys.go [hunk 2]
```diff
@@ -148,41 +148,77 @@ func NewPublicKeyRing() *PublicKeyRing {
...
-// PublicKeyForLocation implements the PublicKeyLocator interface.
+// PublicKeyForLocation implements the PublicKeyLocator interface,
+// by returning the public key most closely associated with loc.
+// If loc is not a valid URL, it returns ErrNotFound; otherwise
+// the host part of the URL must match a registered location.
+//
+// Of those registered locations with matching host parts,
+// longer paths take precedence over short ones.
+// The matching is similar to that of http.ServeMux, except there
+// must be a host part
```
Comment (alesstimec) [inline bakery/keys.go]: full stop (.)

Comment (rogpeppe) [inline bakery/keys.go]: done

**Comment by alesstimec** [comment]
LGTM

**Pull request merged**